### PR TITLE
Java client: receive pubsub messages

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Based on #381 

## Configuration with callback:

```java
    @Test
    @SneakyThrows
    public void basic_client() {
        BiConsumer<Message, Object> callback =
            (msg, context) -> System.out.printf("Received %s, context %s\n", msg, context);

        var regularClient =
                RedisClient.CreateClient(
                                RedisClientConfiguration.builder()
                                        .address(NodeAddress.builder().port(6379).build())
                                        .requestTimeout(3000)
                                        .subscriptionConfiguration(
                                                StandaloneSubscriptionConfiguration.builder()
                                                        .subscription(EXACT, "ch1")
                                                        .subscription(PATTERN, "chat*")
                                                        .callback(callback)
                                                        .build())
                                        .build())
                        .get();

        Thread.sleep(100500);
    }
```

To send messages use `redis-cli`: `publish ch1 $message`

## Configuration without callback:

```java
    @Test
    @SneakyThrows
    public void basic_client() {
        var regularClient =
                RedisClient.CreateClient(
                                RedisClientConfiguration.builder()
                                        .address(NodeAddress.builder().port(6379).build())
                                        .requestTimeout(3000)
                                        .subscriptionConfiguration(
                                                StandaloneSubscriptionConfiguration.builder()
                                                        .subscription(EXACT, "ch1")
                                                        .subscription(PATTERN, "chat*")
                                                        .build())
                                        .build())
                        .get();

        Thread.sleep(100500);

        Message msg = null;
        while ((msg = regularClient.tryGetPubSubMessage()) != null) {
            System.out.printf("Received from queue %s\n", msg);
        }
    }
```

### CODE REVIEW

File: java/client/src/main/java/glide/connectors/handlers/MessageHandler.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            // TODO do we need to create user a message?
+            return;
+        }
+        @SuppressWarnings("unchecked")
+        Map<String, Object> push = (Map<String, Object>) data;
+        PushKind pushType = Enum.valueOf(PushKind.class, (String) push.get("kind"));
+        String[] values = castArray((Object[]) push.get("values"), String.class);
+        switch (pushType) {
+            case Disconnection:
+                // TODO what to do?
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/connectors/handlers/MessageHandler.java:55]: https://github.com/Bit-Quill/glide-for-redis/blob/b0a026b5c4ced0202bb33134bef303f447c86218/python/python/glide/redis_client.py#L292-L299

File: java/client/src/main/java/glide/connectors/handlers/MessageHandler.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            case PSubscribe:
+            case SSubscribe:
+            case Unsubscribe:
+            case PUnsubscribe:
+            case SUnsubscribe:
+                // ignore for now
+                break;
+            default:
+                // TODO log thru error https://github.com/aws/glide-for-redis/pull/1422
+                System.err.printf("Received push with unsupported type: %s.\n", pushType);
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/connectors/handlers/MessageHandler.java:73]: https://github.com/Bit-Quill/glide-for-redis/blob/b0a026b5c4ced0202bb33134bef303f447c86218/python/python/glide/redis_client.py#L323-L330

File: java/src/lib.rs [hunk 3]
```diff
@@ -94,16 +82,62 @@ fn redis_value_to_java<'local>(env: &mut JNIEnv<'local>, val: Value) -> JObject<
...
         }
     }
 }
 
+/// Convert an array of values into java array of corresponding values.
+///
+/// Recursively calls to [`redis_value_to_java`] for every element.
+///
+/// Returns an arbitrary java `Object[]`.
+fn array_to_java_array<'local>(env: &mut JNIEnv<'local>, values: Vec<Value>) -> JObject<'local> {
```
Comment (Yury-Fridlyand) [inline java/src/lib.rs:127]: Moved from `Value::Array` handling from above

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 4]
```diff
@@ -252,31 +283,82 @@ public abstract class BaseClient
...
     }
 
+    /**
+     * Tries to return a next pubsub message.
+     *
+     * @throws WrongConfiguration If client is not subscribed to any channel or if client configured
+     *     with a callback.
+     * @return A message if any or <code>null</code> if there are no unread messages.
+     */
+    public Message tryGetPubSubMessage() {
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:324]: why "try"?  what is the purpose of this method vs `getPubSubMessage`?

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 5]
```diff
@@ -252,31 +283,82 @@ public abstract class BaseClient
...
+        if (subscriptionConfiguration.get().getCallback().isPresent()) {
+            throw new WrongConfiguration(
+                    "The operation will never complete since messages will be passed to the configured"
+                            + " callback.");
+        }
+        return messageQueue.poll();
+    }
+
+    /**
+     * Returns a next pubsub message.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/BaseClient.java:339]: Returns a CompletableFuture

File: java/client/src/main/java/glide/api/models/MessageQueue.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.api.models;
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/MessageQueue.java:2]: why is this in models?

Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/MessageQueue.java:2]: are we going to expose the queue to the end-user?

File: java/src/lib.rs [hunk 7]
```diff
@@ -94,10 +82,62 @@ fn redis_value_to_java<'local>(env: &mut JNIEnv<'local>, val: Value) -> JObject<
             data: _,
             attributes: _,
         } => todo!(),
-        Value::Push { kind: _, data: _ } => todo!(),
+        // Create a java `Map<String, Object>` with two keys:
+        //   - "kind" which responds to the push type, stored as a `String`
+        //   - "values" which responds to the array of values received, stored as `Object[]`
```
Comment (aaron-congo) [inline java/src/lib.rs:87]:
```suggestion
        //   - "kind" which corresponds to the push type, stored as a `String`
        //   - "values" which corresponds to the array of values received, stored as `Object[]`
```

**Review by aaron-congo — APPROVED** [review]

File: java/client/src/main/java/glide/connectors/handlers/ReadHandler.java [hunk 8]
```diff
@@ -31,6 +31,7 @@ public void channelRead(@NonNull ChannelHandlerContext ctx, @NonNull Object msg)
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
         // TODO: log thru logger
         System.out.printf("=== exceptionCaught %s %s %n", ctx, cause);
+        cause.printStackTrace();
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/connectors/handlers/ReadHandler.java:34]: remove?

File: java/client/src/main/java/glide/api/BaseClient.java (hunk 4 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/BaseClient.java:324]: I followed python client namings: https://github.com/aws/glide-for-redis/pull/1616/files#diff-f0f69819afe466c0f3e8afd4ea2fd31ca4f451373fa8ed1351f48473a1c39633R5042
```py
async def get_pubsub_message(self) -> PubSubMsg:
def try_get_pubsub_message(self) -> Optional[PubSubMsg]:
```

File: java/client/src/main/java/glide/connectors/handlers/ReadHandler.java (hunk 8 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/connectors/handlers/ReadHandler.java:34]: I think we need to log it through logger too.

File: java/client/src/main/java/glide/api/models/MessageQueue.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/MessageQueue.java]: I forgot to remove it, thanks.
As discussed, a user should have no queue access.

File: java/client/src/main/java/glide/api/BaseClient.java (hunk 5 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/BaseClient.java]: Updated.

**Pull request merged**